client/src/assets/player/stats/stats-card.ts

```diff
@@ -0,0 +1,184 @@
+  container: HTMLDivElement
+  list: HTMLDivElement
+  closeButton: HTMLElement
+  update: any
+  source: any
+
+  interval = 300
+  playerNetworkInfo: any = {}
+  statsForNerdsEvents = new videojs.EventTarget()
+
+  constructor (player: videojs.Player, options: StatsCardOptions) {
+    super(player, options)
+  }
+
+  createEl () {
+    const container = super.createEl('div', {
+      className: 'vjs-stats-content',
+      innerHTML: getMainTemplate()
+    }) as HTMLDivElement
+    this.container = container
+    this.container.style.display = 'none'
+
+    this.closeButton = this.container.getElementsByClassName('vjs-stats-close')[0] as HTMLElement
+    this.closeButton.onclick = () => this.hide()
+
+    this.list = this.container.getElementsByClassName('vjs-stats-list')[0] as HTMLDivElement
+
+    console.log(this.player_.qualityLevels())
+
+    this.player_.on('p2pInfo', (event: any, data: PlayerNetworkInfo) => {
+      if (!data) return // HTTP fallback
+
+      this.source = data.source
+
+      const p2pStats = data.p2p
+      const httpStats = data.http
+
+      this.playerNetworkInfo.downloadSpeed = bytes(p2pStats.downloadSpeed + httpStats.downloadSpeed).join(' ')
+      this.playerNetworkInfo.uploadSpeed = bytes(p2pStats.uploadSpeed + httpStats.uploadSpeed).join(' ')
+      this.playerNetworkInfo.totalDownloaded = bytes(p2pStats.downloaded + httpStats.downloaded).join(' ')
+      this.playerNetworkInfo.totalUploaded = bytes(p2pStats.uploaded + httpStats.uploaded).join(' ')
+      this.playerNetworkInfo.numPeers = p2pStats.numPeers
+      this.playerNetworkInfo.averageBandwidth = bytes(getAverageBandwidthInStore() || p2pStats.downloaded + httpStats.downloaded).join(' ')
+
+      if (data.source === 'p2p-media-loader') {
+        this.playerNetworkInfo.downloadedFromServer = bytes(httpStats.downloaded).join(' ')
+        this.playerNetworkInfo.downloadedFromPeers = bytes(p2pStats.downloaded).join(' ')
+      }
+    })
+
+    return container
+  }
+
+  toggle () {
+    this.update
+      ? this.hide()
+      : this.show()
+  }
+
+  show (options?: StatsCardOptions) {
+    if (options) this.options_ = options
+
+    let metadata = {}
+
+    this.container.style.display = 'block'
+    this.update = setInterval(async () => {
+      try {
+        if (this.source === 'webtorrent') {
+          const progress = this.player_.webtorrent().getTorrent()?.progress
+          const videoFile = this.player_.webtorrent().getCurrentVideoFile()
+          videoFile.metadata = metadata[videoFile.fileUrl] = videoFile.metadata || metadata[videoFile.fileUrl] || videoFile.metadataUrl && await fetch(videoFile.metadataUrl).then(res => res.json())
+          this.list.innerHTML = getListTemplate(this.options_, this.player_, { playerNetworkInfo: this.playerNetworkInfo, videoFile, progress })
+        } else {
+          this.list.innerHTML = getListTemplate(this.options_, this.player_, { playerNetworkInfo: this.playerNetworkInfo })
+        }
```

Comment:
@Chocobozzz I'm wondering how to get the current progress of buffered data, and more importantly the current video file in the case of `p2p-media-loader` use; I searched through the code, to no avail 😢

Reply:
I think you need to give the ability to provide the `hlsjs` object in the `p2p-media-loader-plugin.ts` so you can get it here and use all available information